[2.0] Fix a bug in Registry.php

Wang-Yu-Chao opened the pull request:
In class Registry, function set() is supposed to return the old value if exists, according to the comment.
But actually return the newly set value, since assignment is right-associated.

Wang-Yu-Chao committed:
[2.0] Fix a bug in Registry.php

mbabker commented on src/Registry.php:
```diff
@@ -508,11 +508,13 @@ public function set($path, $value, $separator = null)
 		{
 			case (is_object($node)):
-				$result = $node->{$nodes[$i]} = $value;
+				$result = $node->{$nodes[$i]};
```
This needs to be `$result = $node->{$nodes[$i]} ?? null;` (similar for the array case), as is it's causing unit test failures because of undefined properties.

Wang-Yu-Chao committed:
[2.0] Fix a bug in Registry.php

Wang-Yu-Chao committed:
[2.0] Fix a bug in Registry.php

mbabker commented on src/Registry.php:
```diff
@@ -508,11 +508,13 @@ public function set($path, $value, $separator = null)
 		{
 			case (is_object($node)):
-				$result = $node->{$nodes[$i]} = $value;
+				$result = $node->{$nodes[$i]} ?? $value;
```
If the intent is to return the previously set value, I'd suggest returning null would be the correct behavior.  Otherwise now you're in a mixed case where if it wasn't previously set you're getting the newly set value and if it was previously set you're getting the old value and that API inconsistency isn't necessarily a good thing.

Wang-Yu-Chao replied:
You are right. But seems that $result = $node->{$nodes[$i]} ?? null cannot pass unit test...strange

mbabker replied:
Which is why I said this had to be fixed on the 2.0 branch only.  Because the current code behavior, and inherently the tests, all return the newly set value and if the intent is to change it to follow the docs and return the old value, the tests need updating to account for that change in behavior.

Wang-Yu-Chao replied:
I see. Thank you for your patience.

Wang-Yu-Chao committed:
[2.0] Fix a bug in Registry.php

Wang-Yu-Chao commented:
Should we correct the code and unit tests here or change the document instead? 
I only find this returned value used in method `setParam` in class `User`, and method `onAjaxSampledataApplyStep1` in class `PlgSampledataBlog`. And seems they don't cause any side effect.

mbabker commented:
Personally, I would suggest to change it.

- [`AbstractApplication::set()`](https://github.com/joomla-framework/application/blob/master/src/AbstractApplication.php#L172) is a wrapper around `Registry::set()` which returns the previous value, it was coded in a way where it bypasses this API inconsistency
- The setters in [`Language`](https://github.com/joomla-framework/language/blob/master/src/Language.php) all return the previous value
- [`Session::set()`](https://github.com/joomla-framework/session/blob/8b8b9519268c9fd979270992631ae04b2a132f76/Session.php#L461) returns the previous value

So our API design favors setters returning the previous values, with that said though there are some places where our setters have no return as well (i.e. our [`Uri`](https://github.com/joomla-framework/uri/blob/master/src/Uri.php) object).  Personally, I favor the return previous value approach, but either way I don't think returning the same value you're setting is correct.

@wilsonge thoughts?

nibra committed:
Merge branch '2.0-dev' into Wang-Yu-Chao_2.0-dev

# Conflicts:
#	src/Registry.php

nibra committed:
Fix unit tests
The tests where looking for a different behaviour than documented. Setting a value in a Registry should actually return the previous value (or null, if set for the first time).

nibra commented:
Thank you!
